### ISSUE TITLE
How to Handle Unsupported Scalar Types

### ISSUE THREAD

**Issue by DexJen** [graphql-dotnet/graphql-dotnet#458]
I need support for byte, short, and double scalar graph types. The system I'm working with uses these types so the GraphQL interface needs to handle them somehow. I can create the `ByteGraphType` class that inherits from `ScalarGraphType`. I can also create a `ByteValue` class. The problem is that `AstFromValue` will throw `ExecutionError($"Cannot convert value to AST: {serialized}")` when it is called. That code has no knowledge of the graph types I've created. There may be other problems I'm not aware of.

I could expose the byte and short properties in the interface as integers and simply check that any input values don't overflow a byte or a short. I'm not sure how I would handle a double though.

I'm confident I could add these scalar types to the code and submit a pull request. But my situation seems to be very unique. I've not found any issues disusing my particular problem so I'm not sure how such a pull request would be received. I'm not sure it would add much value.

So the question I have is; should I change the code and submit a pull request or is there a better way to handle these types?

**Comment by joemcbride** [comment]
> I need support for byte, short, and double scalar graph types

The good news is that the `FloatGraphType` uses `double` internally, so you should be fine there.  The GraphQL spec calls it a `Float`, though the type in .NET that matches the spec is `Double`.

> So the question I have is; should I change the code and submit a pull request or is there a better way to handle these types?

I've had a few other people run into issues with this, so it is probably time to provide a way to hook-in custom handling.  I don't think this should be too difficult to add.  I'm thinking some sort of registration off of `ISchema` that could get ran in `AstFromValue`.  It may be smart to update the `IValue` interface so that each `IValue` can just provide its own value, vs. the framework explicitly checking for each known type in `ValueFromAst`.

Some quick thoughts on the interfaces:

```csharp
public interface IASTFromValueConverter
{
  bool Matches(object value);
  IValue Convert(object value);
}

public interface IValue : INode
{
  object GetValue();
}
```

**Comment by joemcbride** [comment]
```csharp
schema.RegisterValueConverter(new ByteValueConverter())
```

```csharp
public interface IAstFromValueConverter
{
  bool Matches(object value);
  IValue Convert(object value);
}

public interface IValue : INode
{
  object GetValue();
}

public sealed class ByteValue : IValue
{
  public ByteValue(byte value)
  {
    Value = value;
  }

  public byte Value { get; }

  public object GetValue()
  {
    return Value;
  }
}

public sealed class ByteValueConverter : IAstFromValueConverter
{
  public bool Matches(object value)
  {
    return value is byte;
  }

  public IValue Convert(object value)
  {
    return new ByteValue((byte)value);
  }
}

public class ByteGraphType : ScalarGraphType
{
    public ByteGraphType()
    {
        Name = "Byte";
    }

    public override object Serialize(object value)
    {
        return ParseValue(value);
    }

    public override object ParseValue(object value)
    {
        if (value == null)
            return null;

        try
        {
            var result = Convert.ToByte(value);
            return result;
        }
        catch (FormatException)
        {
            return null;
        }
    }

    public override object ParseLiteral(IValue value)
    {
        var byteVal = value as ByteValue;
        return byteVal?.Value;
    }
}
```

**Comment by DexJen** [comment]
I really like this. I'm going move forward with it hopefully submit a pull request.

**Comment by joemcbride** [comment]
@DexJen Got your changes merged, with a few tweaks.  Published to Nuget as [2.0.0-alpha-811](https://github.com/graphql-dotnet/graphql-dotnet/releases/tag/v2.0.0-alpha-811)

```csharp
    internal class ByteValueConverter : IAstFromValueConverter
    {
        public bool Matches(object value, IGraphType type)
        {
            return value is byte;
        }

        public IValue Convert(object value, IGraphType type)
        {
            return new ByteValue((byte)value);
        }
    }
```

**Comment by santoshpatro** [comment]
Hi Joe,

I am using GraphQL 2.3.0 version. Based on your above details I believe I should be able to use the ByteGraphType similar to the other GraphQL.Types. But I am not able to use it. Perhaps I am missing something here.

Can you please help me to get some sample reference of using ByteGraphType. I am trying to return byte[] for a field and seeing an error. 

Any help on this is much appreciated.

Thanks,
Santosh

**Comment by noufionline** [comment]
How can I convert byte[] ?
I have a Timestamp column in SQL Server which is a byte[] in c#. I am not able to convert it. Can you help?


```csharp
 public class ByteGraphType : ScalarGraphType
    {
        public ByteGraphType()
        {
            Name = "ByteArray";
            Description = "ByteGraphType";
        }

        public override object Serialize(object value)
        {
            return ParseValue(value);
        }

        public override object ParseValue(object value)
        {
            byte[] result = Convert.FromBase64String(value.ToString());
            return result;
        }

        public override object ParseLiteral(IValue value)
        {
            if (value is StringValue stringValue)
            {
                return ParseValue(stringValue.Value);
            }

            return null;
        }
    }

```

I am getting the following error


GraphQL.ExecutionError: Cannot convert value to AST: System.Byte[]
 ---> GraphQL.ExecutionError: Cannot convert value to AST: System.Byte[]
   at GraphQL.GraphQLExtensions.AstFromValue(Object value, ISchema schema, IGraphType type)
   at GraphQL.GraphQLExtensions.<>c__DisplayClass17_0.<AstFromValue>b__1(KeyValuePair`2 pair)
   at System.Linq.Enumerable.SelectEnumerableIterator`2.ToList()
   at GraphQL.GraphQLExtensions.AstFromValue(Object value, ISchema schema, IGraphType type)
   at GraphQL.Execution.ExecutionHelper.GetVariableValue(Document document, ISchema schema, VariableDefinition variable, Object input)
   at GraphQL.Execution.ExecutionHelper.GetVariableValues(Document document, ISchema schema, VariableDefinitions variableDefinitions, Inputs inputs)
   at GraphQL.DocumentExecuter.BuildExecutionContext(ISchema schema, Object root, Document document, Operation operation, Inputs inputs, IDictionary`2 userContext, CancellationToken cancellationToken, Metrics metrics, IEnumerable`1 listeners, Boolean throwOnUnhandledException, Action`1 unhandledExceptionDelegate, Nullable`1 maxParallelExecutionCount)
   at GraphQL.DocumentExecuter.ExecuteAsync(ExecutionOptions options)
   --- End of inner exception stack trace ---

**Comment by Emrikol01** [comment]
Hi all,

Was anyone able to get the issue "GraphQL.ExecutionError: Cannot convert value to AST: System.Byte[]" fixed? I have never worked with GraphQL and I've been handed this legacy VS project, (original developer is long gone). The code looks identical to this hereunder but it doesn't work. In my react component I pass a variable: $fileContents: ByteArray! which is of Type File encoded to base64, and it fails.  
 ` public class ByteArrayGraphType : ScalarGraphType
    {
        public ByteArrayGraphType()
        {
            Name = "ByteArray";
        }

        public override object Serialize(object value)
        {
            return ParseValue(value);
        }

        public override object ParseValue(object value)
        {
            if (value == null)
            {
                return null;
            }

            return Parse(value?.ToString());
        }

        public override object ParseLiteral(IValue value)
        {
            if (value?.Value?.ToString() == null)
            {
                return null;
            }

            return Parse(value?.Value?.ToString());
        }
        
        private object Parse(string value)
        {
            try
            {
                var result = Convert.FromBase64String(value?.ToString());
                return result;
            }
           catch (FormatException)
            {
                return null;
            }
        }
    }`

regards, Norman

**Comment by Shane32** [comment]
Custom scalars don't actually work the way they are described in the xml comments, or they way they logically should.  I'm hoping to fix this in 4.0 but it will be a significant rewrite.  Once completed, it will be trivial to write a custom scalar in the manner that you demonstrated.

In the meantime, I'd suggest you use a `StringGraphType` for the input argument type, and use the new built-in behavior to convert the base64 string data to a byte array by simply calling `context.GetArgument<byte[]>("myField")`.  If the string is inside a custom `InputObjectGraphType`, then it might still work, or you might need to convert it to a byte array in your code.
